File: 82_Remove_Duplicates_from_Sorted_list_2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        ans = ListNode()
        res = ans
        while head:
            if head.next and head.val != head.next.val: # if adjacent elements are not same append to res
                ans.next = ListNode(head.val)
                ans = ans.next
                head = head.next
            elif head.next and head.val == head.next.val: # if  adjacent elements are same, skip the current till current head value is skipped
                skip = head.val
                while head and skip == head.val:
                    head = head.next
            else:
                ans.next = ListNode(head.val)
                ans = ans.next
                head = head.next
        
        return res.next

File: 82_Remove_Duplicates_from_Sorted_list_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* newHead = new ListNode(0, head);
        ListNode* newTemp = newHead;
        ListNode* DuplicateTemp = newHead;


        while(DuplicateTemp && DuplicateTemp->next){
    
            DuplicateTemp = DuplicateTemp->next;
            
            while((DuplicateTemp && DuplicateTemp->next) && DuplicateTemp->val == DuplicateTemp->next->val){
                DuplicateTemp = DuplicateTemp->next;
                
                if(DuplicateTemp->next == NULL || DuplicateTemp->val != DuplicateTemp->next->val)
                    DuplicateTemp = DuplicateTemp->next;
            }
            
            newTemp->next = DuplicateTemp;
            newTemp = newTemp->next;
        }
        
        return newHead->next;
    }
};